lib/ansible/utils/encrypt.py

```diff
@@ -33,31 +40,140 @@
 DEFAULT_PASSWORD_LENGTH = 20
 
 
-def do_encrypt(result, encrypt, salt_size=None, salt=None):
-    if PASSLIB_AVAILABLE:
+class BaseHash(object):
+    algo = namedtuple('algo', ['crypt_id', 'salt_size', 'implicit_rounds'])
+    algorithms = {
+        'md5_crypt': algo(crypt_id='1', salt_size=8, implicit_rounds=None),
+        'bcrypt': algo(crypt_id='2a', salt_size=22, implicit_rounds=None),
+        'sha256_crypt': algo(crypt_id='5', salt_size=16, implicit_rounds=5000),
+        'sha512_crypt': algo(crypt_id='6', salt_size=16, implicit_rounds=5000),
+    }
+
+    def __init__(self, algorithm, error_type):
+        self.algorithm = algorithm
+        self.error_type = error_type
+
+
+class CryptHash(BaseHash):
+    def __init__(self, algorithm, error_type):
+        super(CryptHash, self).__init__(algorithm, error_type)
+
+        if sys.platform.startswith('darwin'):
+            raise self.error_type("crypt.crypt not supported on Mac OS X/Darwin, install passlib python module")
+
+        if algorithm not in self.algorithms:
+            raise self.error_type("crypt.crypt does not support '%s' algorithm" % self.algorithm)
+        self.algo_data = self.algorithms[algorithm]
+
+    def hash(self, secret, **settings):
+        salt = self._salt(**settings)
+        rounds = self._rounds(**settings)
+        return self._encrypt(secret, salt, rounds)
+
+    def _salt(self, **settings):
+        salt_size = settings.get('salt_size') or self.algo_data.salt_size
+        return settings.get('salt') or random_salt(salt_size)
+
+    def _rounds(self, **settings):
+        if settings.get('rounds') == self.algo_data.implicit_rounds:
+            # Passlib does not include the rounds if it is the same as implict_rounds.
+            # Make crypt lib behave the same, by not explicitly specifying the rounds in that case.
+            return None
+        return settings.get('rounds')
+
+    def _encrypt(self, secret, salt, rounds):
```

Comment:
Don't use the name _encrypt for hashing... I know we already have a publically available function named do_encrypt which really does hashing but there's no sense making that problem worse.